lib/srv/usermgmt.go

```diff
@@ -389,22 +315,50 @@ func (u *HostUserManagement) UpsertUser(name string, ui *services.HostUsersInfo)
 		}
 
 		return nil
-	})
-	if err != nil {
-		return nil, trace.Wrap(err)
+	}))
+}
+
+// UpsertUser creates a temporary Teleport user in the TeleportServiceGroup
+func (u *HostUserManagement) UpsertUser(name string, ui *services.HostUsersInfo) (io.Closer, error) {
+	// grabbing a copy so we don't mutate the caller's HostUsersInfo
+	uiWithGroups := *ui
+	uiWithGroups.Groups = make([]string, 0, len(ui.Groups))
+
+	var groupErrs []error
+	for _, group := range append(ui.Groups, types.TeleportServiceGroup) {
+		if err := u.createGroupIfNotExist(group); err != nil {
+			groupErrs = append(groupErrs, err)
+		}
+
+		if group != types.TeleportServiceGroup {
+			uiWithGroups.Groups = append(uiWithGroups.Groups, group)
+		}
+	}
+
+	if err := trace.NewAggregate(groupErrs...); err != nil {
+		return nil, trace.WrapWithMessage(err, "error while creating groups")
 	}
 
-	if ui.Mode == types.CreateHostUserMode_HOST_USER_MODE_KEEP {
-		return nil, nil
+	var closer *userCloser
```

Comment:
I'm getting a weird panic when a user exits a session in `keep` mode (like `nil != nil == true` weird). This seems to fix it 🤷.
```suggestion
	var closer io.Closer
```

Reply:
It looks like `assert.Nil()` (and by proxy, `assert.NotNil()`) go to some lengths to support treating nil interfaces as nil 🤔 Swapping those with `assert.Equal(t, nil, closer)` and `assert.NotEqual(t, nil, closer` seems to catch the bad behavior